src/main/java/com/openlattice/datastore/services/BackgroundIndexingService.java

```diff
@@ -0,0 +1,147 @@
+        this.elasticsearchApi = elasticsearchApi;
+
+        this.propertyTypes = hazelcastInstance.getMap( HazelcastMap.PROPERTY_TYPES.name() );
+        this.entityTypes = hazelcastInstance.getMap( HazelcastMap.ENTITY_TYPES.name() );
+        this.entitySets = hazelcastInstance.getMap( HazelcastMap.ENTITY_SETS.name() );
+
+        indexUpdatedEntitySets();
+
+    }
+
+    private String getDirtyEntitiesQuery( UUID entitySetId ) {
+        return "SELECT * FROM " + DataTables.entityTableName( entitySetId ) + " WHERE "
+                + PostgresColumn.LAST_INDEX_FIELD + " < " + PostgresColumn.LAST_WRITE_FIELD + " LIMIT " + String
+                .valueOf( BLOCK_SIZE )
+                + " OFFSET ?";
+    }
+
+    private Set<UUID> getDirtyEntityKeyIds( UUID entitySetId, long offset ) {
+        String entitySetTableName = DataTables.entityTableName( entitySetId );
+
+        try ( Connection connection = hds.getConnection();
+                PreparedStatement ps = connection.prepareStatement( getDirtyEntitiesQuery( entitySetId ) ) ) {
+            ps.setString( 1, entitySetTableName );
+            ps.setLong( 2, offset );
+
+            ResultSet rs = ps.executeQuery();
+
+            Set<UUID> result = Sets.newHashSet();
+            while ( rs.next() && result.size() < BLOCK_SIZE ) {
+                result.add( ResultSetAdapters.id( rs ) );
+            }
+            rs.close();
+            return result;
+        } catch ( SQLException e ) {
+            logger.debug( "Unable to load all entity set ids", e );
+            return ImmutableSet.of();
+        }
+    }
+
+    private Map<UUID, Map<UUID, PropertyType>> getPropertyTypesByEntityTypesById() {
+        Map<UUID, Map<UUID, PropertyType>> result = Maps.newHashMap();
+
+        entityTypes.entrySet().forEach( entry -> {
+            UUID entityTypeId = entry.getKey();
+            Map<UUID, PropertyType> propertyTypesById = propertyTypes
+                    .getAll( entityTypes.get( entityTypeId ).getProperties() ).entrySet().stream()
+                    .filter( ptEntry -> !ptEntry.getValue().getDatatype().equals( EdmPrimitiveTypeKind.Binary ) )
+                    .collect(
+                            Collectors.toMap( ptEntry -> ptEntry.getKey(), ptEntry -> ptEntry.getValue() ) );
+            result.put( entityTypeId, propertyTypesById );
+        } );
+
+        return result;
+    }
+
+    @Scheduled( fixedDelay = 500 )
+    public void indexUpdatedEntitySets() {
+
+        Map<UUID, Map<UUID, PropertyType>> propertyTypesByEntityType = getPropertyTypesByEntityTypesById();
+
+        entitySets.values().forEach( entitySet -> {
+            if ( !entitySet.getName().equals( AuditEntitySetUtils.AUDIT_ENTITY_SET_NAME ) ) {
+                UUID entitySetId = entitySet.getId();
+                Map<UUID, PropertyType> propertyTypeMap = propertyTypesByEntityType
+                        .get( entitySet.getEntityTypeId() );
+
+                long offset = 0;
```

Comment:
I see now. You are using offset to implement cursor behavior.  You should actually set fetchSize on the prepared statement and let postgres automatically handle the buffering as you scan out a number of fixed size chunks. 

More details here: https://jdbc.postgresql.org/documentation/head/query.html#query-with-cursor